### PR TITLE
abstract pubsub service via gocloud

### PR DESCRIPTION
**Pull request by pxp928**
# Description of the PR

Utilizing [gocloud](https://gocloud.dev/howto/pubsub/) we can abstract out the pubsub used within GUAC. By default, it does still work with NATS Jetstream but now has the ability for other cloud provider/third-party pubsub services to be used (such as AWS, GCP, Azure. NATS, Kafka and RabbitMQ.

I tested it with AWS S3 as the blob store and AWS SQS as the pubsub with authentication handled via AWS CLI:

```
go run ./cmd/guacingest --blob-addr "s3://test?region=us-east-2" --pubsub-addr "awssqs://sqs.us-east-2.amazonaws.com/123456789012/myqueue"?region=us-east-2"
```

and in another terminal:

```
go run ./cmd/guaccollect files --blob-addr "s3://test?region=us-east-2" --pubsub-addr "awssqs://sqs.us-east-2.amazonaws.com/123456789012/myqueue"?region=us-east-2" ../guac-data/docs
```

Also fixes https://github.com/guacsec/guac/issues/765

Also addressed issue https://github.com/guacsec/guac/issues/460 with the addition of sending CDEvents via the pubsub

# PR Checklist

- [x] All commits have [a Developer Certificate of Origin (DCO)](https://wiki.linuxfoundation.org/dco) -- they are generated using `-s` flag to `git commit`.
- [x] All new changes are covered by tests
- [x] If GraphQL schema is changed, `make generate` has been run
- [x] If `collectsub` protobuf has been changed, `make proto` has been run
- [x] All CI checks are passing (tests and formatting)
- [x] All dependent PRs have already been merged

### CODE REVIEW

File: pkg/emitter/emitter.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+		return fmt.Sprintf("%s?subject=%s", serviceURL, subjectNameDocCollected)
+	} else {
+		return serviceURL
+	}
+}
+
+// buildSubscriptionURL constructs the full URL for subscription.
+// If using NATS, additional parameters are needed for jetstream
+func buildSubscriptionURL(serviceURL string) string {
+	if strings.Contains(serviceURL, "nats://") {
```
Comment (jeffmendoza) [inline pkg/emitter/emitter.go:76]: Nit: All of the `strings.Contains` should probably be switched with https://pkg.go.dev/strings#HasPrefix

File: pkg/certifier/certify/certify.go [hunk 2]
```diff
@@ -164,15 +166,34 @@ func generateDocuments(ctx context.Context, collectedComponent interface{}, emit
...
+
+	keyByte, err := json.Marshal(cdEvent)
+	if err != nil {
+		return fmt.Errorf("failed marshal of document key: %w", err)
+	}
+
+	if err := pubsub.Publish(ctx, keyByte); err != nil {
+		if err != nil {
+			return fmt.Errorf("failed to publish event with error: %w", err)
+		}
```
Comment (jeffmendoza) [inline pkg/certifier/certify/certify.go:196]: Are these updates for the nats->gocloud change or for the blobstore?

File: pkg/cli/store.go [hunk 3]
```diff
@@ -35,7 +35,7 @@ func init() {
 
 	// Set of all flags used across GUAC clis and subcommands. Use consistent
 	// names for config file.
-	set.String("nats-addr", "nats://127.0.0.1:4222", "address to connect to NATs Server")
+	set.String("pubsub-addr", "nats://127.0.0.1:4222", "address to connect to a pubsub (default is NATS)")
```
Comment (jeffmendoza) [inline pkg/cli/store.go:38]: Instead of "address" maybe "gocloud connection string"?

**Review by jeffmendoza — APPROVED** [review]
Should figure out the TODO and commented out code before merging probably.

File: pkg/certifier/certify/certify.go (hunk 2 above)
Comment (pxp928) [reply pkg/certifier/certify/certify.go:196]: The certifier was missed when the blobstore was added. So this updates the publish for blobstore and pubsub

File: pkg/emitter/emitter.go (hunk 1 above)
Comment (pxp928) [reply pkg/emitter/emitter.go:76]: Done!

File: pkg/cli/store.go (hunk 3 above)
Comment (pxp928) [reply pkg/cli/store.go:38]: fixed

**Comment by pxp928** [comment]
> Should figure out the TODO and commented out code before merging probably.

Removed the commented out code as we are not using it.

**Review by mihaimaruseac — APPROVED** [review]

**Pull request merged**